fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -874,6 +876,104 @@ ACTOR Future<Void> preresolutionProcessing(CommitBatchContext* self) {
+                                                                             MutationRef m) {
+	std::pair<EncryptCipherDomainName, int64_t> details(EncryptCipherDomainName(), TenantInfo::INVALID_TENANT);
+
+	// Possible scenarios:
+	// 1. Encryption domain (Tenant details) weren't explicitly provided, extract Tenant details using
+	// TenantPrefix (first 8 bytes of FDBKey)
+	// 2. Encryption domain isn't available, leverage 'default encrypiton domain'
+	bool useDefaultEncryptionDomain = true;
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		if (isSystemKey(m.param1)) {
+			// Encryption domain == FDB SystemKeyspace encryption domain
+			details.first = EncryptCipherDomainName(FDB_DEFAULT_ENCRYPT_DOMAIN_NAME);
+			details.second = SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID;
+		} else if ((MutationRef::Type)m.type == MutationRef::Type::ClearRange) {
+			// FIXME: Handle Clear-range transaction, actions needed:
+			// 1. Transaction range can spawn multiple encryption domains (tenants)
+			// 2. Transaction can be a multi-key transaction spawning multiple tenants
+			// For now fallback to 'default encryption domain'
+			ASSERT(useDefaultEncryptionDomain);
+
+			CODE_PROBE(true, "SingleKey ClearRange mutation encryption");
+		} else if (m.param1.size() >= TENANT_PREFIX_SIZE) {
+			// Parse mutation key to determine mutation encryption domain
+			StringRef prefix = m.param1.substr(0, TENANT_PREFIX_SIZE);
+			int64_t tenantId = TenantMapEntry::prefixToId(prefix, false /* enforceValidTenantId */);
+
+			if (tenantId != TenantInfo::INVALID_TENANT) {
+				ErrorOr<Optional<std::pair<TenantName, int64_t>>> result =
+				    getTenantDetails(commitData, tenantId, false);
+				if (!result.isError() && result.get().present()) {
+					details.first = result.get().get().first;
+					details.second = result.get().get().second;
+					useDefaultEncryptionDomain = false;
+				}
+			} else if (commitData->db->get().client.tenantMode == TenantMode::REQUIRED) {
+				// Disallow non-tenant non-clearrange transaction if cluster is configured with 'TenantMode::Required'
+				TraceEvent(SevError, "InvalidMutation", commitData->dbgid)
+				    .detail("MutationType", (MutationRef::Type)m.type);
+				throw encrypt_invalid_transaction();
+			} else {
+				// Leverage 'default encryption domain'
+			}
+		}
+	} else {
+		// ClearRange is the 'only' MultiKey transaction allowed
+		if (m.type != MutationRef::Type::ClearRange) {
+			TraceEvent(SevError, "InvalidMutation", commitData->dbgid)
+			    .detail("MutationType", (MutationRef::Type)m.type);
+			throw encrypt_invalid_transaction();
+		}
+
+		// FIXME: Handle Clear-range transaction, actions needed:
+		// 1. Transaction range can spawn multiple encryption domains (tenants)
+		// 2. Transaction can be a multi-key transaction spawning multiple tenants
+		// For now fallback to 'default encryption domain'
+		ASSERT(useDefaultEncryptionDomain);
+
+		CODE_PROBE(true, "MultiKey ClearRange mutation encryption");
```

Comment:
In tests do we have clear range with end keys contain user data? If not, can we add assert for it?